File: 백준/Silver/2578. 빙고/빙고.py
import sys


def input():
    return sys.stdin.readline().rstrip()


def solution():
    board = [list(map(int, input().split())) for _ in range(5)]
    numbers = [list(map(int, input().split())) for _ in range(5)]
    number_dict = {board[i][j]: (i, j) for i in range(5) for j in range(5)}

    def check_bingo():
        bingo = 0
        # 가로, 세로
        for i in range(5):
            if all(board[i][j] == 0 for j in range(5)):
                bingo += 1
            if all(board[j][i] == 0 for j in range(5)):
                bingo += 1
        # 대각선
        if all(board[i][i] == 0 for i in range(5)):
            bingo += 1
        if all(board[i][4 - i] == 0 for i in range(5)):
            bingo += 1
        return bingo

    cnt = 0

    for i in range(5):
        for j in range(5):
            cnt += 1
            x, y = number_dict[numbers[i][j]]
            board[x][y] = 0

            if check_bingo() >= 3:
                return cnt


print(solution())

File: 백준/Silver/2578. 빙고/빙고.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const readline = require("readline");
let line = 0;
const lines = [];
const input = () => lines[line++];

function readFile(filePath) {
  const readStream = fs.createReadStream(filePath);
  const rl = readline.createInterface({
    input: readStream,
  });

  rl.on("line", (line) => {
    lines.push(line);
  }).on("close", () => {
    console.log(solution());
    process.exit();
  });
}

readFile(filePath);

function solution() {
  const TARGET_BINGO_COUNT = 3;

  const n = 5;
  const board = Array.from({ length: n }, () => input().split(" ").map(Number));
  const numberPositions = new Map();
  board.forEach((row, i) => {
    row.forEach((num, j) => {
      numberPositions.set(num, [i, j]);
    });
  });

  const rowCounts = Array(n).fill(0);
  const colCounts = Array(n).fill(0);
  let diag1Count = 0,
    diag2Count = 0;
  let bingoCount = 0;

  const calledNumbers = Array.from({ length: n }, () =>
    input().split(" ").map(Number)
  ).flat();

  for (let i = 0; i < calledNumbers.length; i++) {
    const num = calledNumbers[i];
    if (numberPositions.has(num)) {
      const [row, col] = numberPositions.get(num);
      updateBingo(row, col);

      if (bingoCount >= TARGET_BINGO_COUNT) return i + 1;
    }
  }

  return "No bingo";

  function updateBingo(row, col) {
    if (++rowCounts[row] === n) bingoCount++;
    if (++colCounts[col] === n) bingoCount++;
    if (row === col && ++diag1Count === n) bingoCount++;
    if (row + col === n - 1 && ++diag2Count === n) bingoCount++;
  }
}
